File: 누적 합 부수기/BOJ_11660_구간_합_구하기_5.java
package BOJ_11660;

import java.io.*;
import java.util.StringTokenizer;

public class BOJ_11660_구간_합_구하기_5 {
    // input
    private static BufferedReader br;

    // variables
    private static int N, M;
    private static int[][] board;

    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\Java Algorithm\\JavaAlgorithm\\src\\BOJ_11660\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() throws IOException {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < M; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            sb.append(
                    prefixSum(
                            Integer.parseInt(st.nextToken()),
                            Integer.parseInt(st.nextToken()),
                            Integer.parseInt(st.nextToken()),
                            Integer.parseInt(st.nextToken())
                    )
            ).append('\n');
        }

        return sb.toString();
    } // End of solve

    private static int prefixSum(int x1, int y1, int x2, int y2) {
        return board[x2][y2] - board[x1 - 1][y2] - board[x2][y1 - 1] + board[x1 - 1][y1 - 1];
    } // End of prefixSum()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N + 1][N + 1];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                board[i + 1][j + 1] = Integer.parseInt(st.nextToken());
                // 대각선으로 교차 합
                board[i + 1][j + 1] = board[i + 1][j] + board[i][j + 1] - board[i][j] + board[i + 1][j + 1];
            }
        }
    } // End of input()
} // End of Main class

File: 누적 합 부수기/BOJ 11660 구간 합 구하기 5.kt
package BOJ_11660

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.OutputStreamWriter
import java.util.*

// input
private lateinit var br: BufferedReader

// variables
private var N = 0
private var M = 0

private lateinit var board: Array<IntArray>

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_11660\\res.txt"
    br = BufferedReader(File(path).bufferedReader())
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    input()

    bw.write(solve())
    bw.close()
} // End of main

private fun solve(): String {
    val sb = StringBuilder()

    StringTokenizer(br.readLine()).also {
        println(it)
        println(it.nextToken())
    }


    repeat(M) {
        StringTokenizer(br.readLine()).run {
            val r1 = nextToken().toInt()
            val c1 = nextToken().toInt()
            val r2 = nextToken().toInt()
            val c2 = nextToken().toInt()


        }
    }

    return sb.toString()
} // End of solve

private fun input() {
    StringTokenizer(br.readLine()).run {
        N = nextToken().toInt()
        M = nextToken().toInt()
    }

    board = Array(N) { IntArray(N) }

    for (i in 0 until N) {
        StringTokenizer(br.readLine()).run {
            for (j in 0 until N) {
                board[i][j] = nextToken().toInt()
            }
        }
    }
} // End of input